### ISSUE TITLE
[NFR] ORM. Ability to set custom Resultset class returned by find()

### ISSUE THREAD

**Issue by ashpumpkin** [phalcon/cphalcon#12166]
Hello. I think this feature is a good idea in cases when there`s some logic related to sets of models
Raugh but simple usecase example:
Day statistics table with columns (date, statistics_type, value).  statistics_type can be 1,2 or 3.
You need to render a month graph with 3 types of stats on it, and inform user about min and max values

``` PHP
// you get Phalcon\Mvc\Model\Resultset\Simple
$stats = Stats::find([
    'date between "2016-06-01" AND "2016-06-30"',
    'order' => 'date'
]);

$graphData = [1 => [], 2 => [], 3 => []];
$maxValue  = $stats[0]->value;
$minValue  = $stats[0]->value;

foreach ($stats as $item) {
    $graphData[$item->statistics_type][] = [$item->date, $item->value];

    if ($item->value < $minValue) {
        $minValue = $item->value;
    }

    if ($item->value > $maxValue) {
        $maxValue = $item->value;
    }
}

$this->view->graphData = $graphData;
$this->view->maxValue  = $maxValue;
$this->view->minValue  = $minValue;
```

The view will take this data and render a page for current user. But the problem is that there are views and other objects where this data is fully or partially needed. And you need to always duplicate the code above or to seek for a place where this logic can be stored. I think Custom resultsets would be a good place for this.

``` PHP
namespace Models;
class Stats extends \Phalcon\Mvc\Model {
    /** CODE */
    public function getResultsetClass() {
        return 'Resultsets\Stats';
    }
    /** CODE */
}
```

``` PHP
namespace Resultsets;
class Stats extends \Phalcon\Mvc\Model\Resultset\Simple {
    public function getGraphData() {
        /** CODE */
    }

    public function getMaxValue() {
        /** CODE */
    }

    public function getMinValue() {
        /** CODE */
    }
}
```

``` PHP
// you get Resultsets\Stats
$stats = Stats::find([
    'date between "2016-06-01" AND "2016-06-30"',
    'order' => 'date'
]);

$this->view->graphData = $stats->getGraphData();
$this->view->minValue  = $stats->getMinValue();
$this->view->maxValue  = $stats->getMaxValue();

// OR $this->view->stats = $stats; and the view will iteract with $stats by itself
```

Also it will help contributors to create their own Resultset classes for incubator.

**Comment by Jurigag** [comment]
Cool idea.

**Comment by tugrul** [comment]
Also able to make batch operations like change status.

**Comment by Jurigag** [comment]
I think you can already do some batch things with resultset like:

```php
$resultset->update([
    'status' => 2
]);
```

Will update every row in resultset.

**Comment by tugrul** [comment]
@Jurigag thanks. this is enough for atomic operations but this can't change easily ongoing revisions if you have this snippet many place in big codebase.

Change single function ease than multiple places.

**Comment by Jurigag** [comment]
@tugrul well i guess there will be no problem at all to add method like `updateStatuses()` then to this resultset class and calls it :)

**Comment by tugrul** [comment]
```php

namespace App\Models\Content;

class Category extends \Phalcon\Mvc\Model
{
    public $id;
    public $name;
    public $parent;
    
    public function getSource()
    {
        return 'content_categories';
    }

    public function initialize()
    {
        $this->hasMany('id', 'App\Models\Content\Category', 'parent', ['alias' => 'Childs']);
        $this->belongsTo('parent', 'App\Models\Content\Category', 'id', ['alias' => 'Parent']);
    }

}

namespace Resultsets;
class Categories extends \Phalcon\Mvc\Model\Resultset\Simple {
    public function executeRecursive($callback, $level = 0) {
		
		if (!is_callable($callback)) {
			return;
		}
		
		foreach ($this as $category) {
			$callback($category, $level);
			$category->getChilds()->executeRecursive($callback, $level + 1);
		}
		
	}
}

echo '<select name="category">';
App\Models\Content\Category::find(['conditions' => 'parent IS NULL'])->executeRecursive(function($category, $level){
	echo '<option value="', $category->id ,'">', str_repeat('&ndash; ', $level), $category->name, '</option>';
});
echo '</select>';


```

**Comment by Fenikkusu** [comment]
As a temporary solution to this, you are able to override the static find method on the model in question and force it to return the resultset you wish (return new Resultset('model', parent::find(...)); or something to that effect. I forget the exact logic I've used to acomplish this. It's hacky but gets the trick done.

**Comment by ashpumpkin** [comment]
https://github.com/phalcon/cphalcon/pull/12359

**Comment by ashpumpkin** [comment]
Implemented in `3.1.x`

**Comment by virgofx** [comment]
@ashpumpkin Can this be revised to only go through the autoloader once -- instead of using class_implements() load the class instance and then check instanceof which will be faster as the autoloading() process only gets hit once.
